Add viscoelasticity to the visco plastic material model

naliboff opened the pull request:
This PR adds the option to use viscoelasticity in the visco plastic material model. Either before the PR is merged or afterward, the viscoelastic plastic material can be removed as all of its functionality will be covered in the visco plastic model. 

At the moment, this is a work in progress and I am opening a PR only for initial review and testing. Note that the documentation in the compute_isostrain_viscosities function has been rewritten with the hope that the order of operations is clearer.
* [X] I have followed the [instructions for indenting my code](../CONTRIBUTING.md#making-aspect-better).
* [X] I have tested my new feature locally to ensure it is correct.
* [X] I have [created a testcase](http://www.math.clemson.edu/~heister/manual.pdf#sec%3Awriting_tests) for the new feature/benchmark in the [tests/](../tests/) directory.
* [X] I have added a changelog entry in the [doc/modules/changes](../doc/modules/changes) directory that will inform other users of my change.

naliboff commented:
Note that this PR is derived from an early version of #3292 (see first commit) and will be updated once that PR is merged.

naliboff commented:
@tjhei - I think this is ready for a review if you have time to take a look. My preference would be to merge this PR and then rebase #3119 so we can test it with strain-softening and elasticity.

tjhei commented on source/material_model/visco_plastic.cc:
```diff
@@ -140,19 +140,58 @@ namespace aspect
           const double viscosity_diffusion = diffusion_creep.compute_viscosity(pressure, temperature_for_viscosity, j);
 
-          // Compute visocisty from dislocation creep law
+          // Step 1b: Compute visocisty from dislocation creep law
```
```suggestion
          // Step 1b: Compute viscosity from dislocation creep law
```

tjhei commented on source/material_model/rheology/elasticity.cc:
```diff
@@ -180,10 +180,10 @@ namespace aspect
-                    ExcMessage("Material models with elasticity cannot be used with "
-                               "material averaging. The variable 'Material averaging' "
-                               "in the 'Material model' subsection must be set to 'none'."));
+        //AssertThrow(this->get_parameters().material_averaging == MaterialModel::MaterialAveraging::none,
```
what do we want to do about this? Can we remove this change and tackle this separately?

naliboff replied:
Yes, lets tackle this separately. None of the tests use cell averaging, so nothing should break by re-introducing this exception.

tjhei commented on source/material_model/visco_plastic.cc:
```diff
@@ -178,49 +217,81 @@ namespace aspect
+          double current_stress;
+          if (use_elasticity == false)
+            {
+              current_edot_ii = edot_ii;
```
I would assign these values unconditionally (two lines up)

tjhei commented on source/material_model/visco_plastic.cc:
```diff
@@ -178,49 +217,81 @@ namespace aspect
+              current_edot_ii = edot_ii;
+              current_stress = viscous_stress;
+            }
+          else
```
and only keep this as ``if (use_elasticity)``

tjhei commented on source/material_model/visco_plastic.cc:
```diff
@@ -593,6 +693,12 @@ namespace aspect
                              "0.3 $K/km$ = 0.0003 $K/m$ = 9.24e-09 $K/Pa$ gives an earth-like adiabat."
                              "Units: $K/Pa$");
+
+          prm.declare_entry ("Include viscoelasticity","false",
```
```suggestion
          prm.declare_entry ("Include viscoelasticity", "false",
```

tjhei commented on source/material_model/visco_plastic.cc:
```diff
@@ -593,6 +693,12 @@ namespace aspect
+
+          prm.declare_entry ("Include viscoelasticity","false",
+                             Patterns::Bool (),
+                             "Whether to inlcude viscoelasticity in the rheological formulation. ");
```
```suggestion
                             "Whether to include viscoelasticity in the rheological formulation.");
```

tjhei commented on source/material_model/visco_plastic.cc:
```diff
@@ -815,6 +929,107 @@ namespace aspect
+                                   "viscoelastic deformation, thus producing a viscoelastic plastic constitutive "
+                                   "relationship. "
+                                   "\n\n "
+                                   "The viscoelastic rheology behavior takes into account the elastic shear "
```
Independent of this PR, but can I get you to write a section in the manual with these formulas? Something similar to section 2.12 or 2.13. This might be a better location for references and this lengthy description.

naliboff replied:
Yes, will do. I've opened an issue (#3365) and assigned myself.

tjhei commented on source/material_model/visco_plastic.cc:
```diff
@@ -140,19 +140,58 @@ namespace aspect
+
+      std::vector<double> elastic_shear_moduli(volume_fractions.size());
+
+      std::vector<double> stresses_ve(volume_fractions.size());
```
Can you fill these variables with signaling NaNs?

tjhei commented:
> My preference would be to merge this PR and then rebase #3119 so we can test it with strain-softening and elasticity.

I think we can tackle these independently as well: I will prepare a simple version of the prescribed terms (without the material model updates).

naliboff commented:
@tjhei - Thanks for the review and I believe I've addressed all of the comments.

Regarding the correctness of the implementation relative to other material models, all of the new test results were copied directly from existing tests that either used the `viscoelastic_plastic` or `viscoelastic` material model. 

Only two of six test results changed (see `update test results` commit) and the changes are fairly minor. The ordering of how compositional fields are averaged is slightly different between material models, and I'm actually surprised most of the tests passed without modification.

@dansand - are you satisfied with the current structure and are the new benchmarks behaving roughly as expected? 

Reviews from any other frequent users of `visco_plastic` would be helpful, even if only to comment on the revised documentation.

naliboff commented:
@gassmoeller - I'm in the process of updating some of the benchmarks to use `visco_plastic` instead of `viscoelastic_plastic`. In doing this, it will actually break some of the` viscoelastic_plastic` tests tied to the benchmarks.

Two options for how to proceed:

1.  Remove the viscoelastic_plastic material model and associated tests in the same commit.
2. Update the viscoelastic_plastic test parameter files and then remove everything in a follow-up PR.

Preferences for one of these two options or some other solution? No strong preference on my side.

gassmoeller commented:
What about option 3: Update the test results according to the new benchmark results? I would like to keep the tests, and I assume that while you update the benchmarks to the visco_plastic model you check that the new (slightly changed) results are still correct?

naliboff commented:
@gassmoeller - that seems like a good solution.

> I assume that while you update the benchmarks to the visco_plastic model you check that the new (slightly changed) results are still correct?

For the most part I've been running all the existing benchmarks with every new commit that changes source code, but I just checked on one of the benchmarks again last night and found some spurious behavior.

My famous last words from above
> Only two of six test results changed (see update test results commit) and the changes are fairly minor
were clearly ill-informed ;)

I'll check back in when I'm satisfied with all the benchmarks and the test suites have been updated accordingly.

naliboff commented:
Good news, the spurious benchmark result was the result unphysical artifacts in the viscoelastic stress field arising from fixed compositional boundary conditions. Removing these boundary conditions produces a perfect fit to the benchmark results. The order of averaging in `viscoelastic_plastic` somehow manages to prevent propagation of the edge effects observed when using` visco_plastic` for the same problem.

@gassmoeller @tjhei - I'm going to prepare a series pull requests that update tests and benchmarks for `viscoelastic` and `viscoelastic_plastic`. These changes will help simplify this PR and make it much clearer what test results change between material models.

naliboff commented:
FYI, I re-ran one of the full VEP benchmarks earlier today and am unfortunately getting a deviation from the analytical results. I've traced this back to a rebase, but I'm going to need to spend a bit of time figuring out exactly what the issue is. In any case, this will be a good opportunity to go back and redo how the new tests are added, which was not done in an ideal manner in the last few commits.

naliboff commented:
Resolved the issue and the aforementioned benchmarks are now passing. See line 260-263 in `visco_plastic.cc` for an explanation of the issue, which was related to differences between how viscous and viscoelastic-plastic stresses are calculated and treated in the `Drucker Prager` rheology module.

In addition to the fix, I've also significantly cleaned up the commit history and added more documentation. @dansand and I are discussing one additional vep benchmark, but the current structure should be by and large complete. 

@tjhei - As soon as this is merged, I think we can safely move on with the associated plasticity testing and PRs. Sorry that this has held up those PRs and related testing!

naliboff commented:
@gassmoeller - This is now ready for review, as all the relevant tests and benchmarks have been modified or added. In principle, we should now be able to delete `viscoelastic_plastic`.

naliboff commented:
@gassmoeller - One point I forgot to add above is that I am not happy with how unruly and extensive the documentation for this material model now is. I think quite a bit can be removed or moved into the relevant rheology module sections, and more in-depth explanations added in the manual. #3365 is a reminder for this task.

naliboff commented on source/material_model/visco_plastic.cc:
```diff
@@ -178,49 +216,80 @@ namespace aspect
               {
+                //Step 4c-1: always rescale the viscosity back to the yield surface
                 const double viscosity_limiter = yield_stress / (2.0 * ref_strain_rate)
                                                  * std::pow((edot_ii/ref_strain_rate), 1./exponents_stress_limiter[j] - 1.0);
```
@gassmoeller - hmmm, ood question. I have not tested the stress limiter with elasticity and I'm not sure there is a benchmark I can compare with. @cedrict and I tested this for simple visco-plastic shear band problems originally, and perhaps the best thing to do is enforce that the VEP rheology is only used in combination with the `Drucker Prager` rheology, module for which we have benchmarks.

naliboff commented on source/material_model/visco_plastic.cc:
```diff
@@ -178,49 +216,80 @@ namespace aspect
+              // The viscoelastic strain rate is divided by 2 here as the Drucker Prager
+              // viscosity calculation assumes stress = 2 * viscosity * strain_rate_invariant,
+              // whereas the combined the viscoelastic + viscous stresses do not include the
+              // 2x factor.
```
@gassmoeller - We absolutely could do `stresses_ve[j] = 2. * viscosities_ve[j] * edot_ii_ve` and not modify the strain rate, but then this affects the stress in line 284 (prior to your commit):
  `if (current_stress >= yield_stress)`

So, I think adding the factor of 2 in at line 33 would means are stresses would be 2x higher than they are currently defined here and in `viscoelastic_plastic`. However, I may be wrong about this and did not just read through the code logic correctly (definitely easier to read after your commit 👍  ).

naliboff replied:
Regarding the stress limiter, per the comment above I think we should just restrict that yielding option to only work with a visco plastic rheology (i.e., without elasticity).

naliboff commented:
@gassmoeller - thanks for the review and commit! See the responses above in regards to the remaining issues.

gassmoeller commented:
Sorry, I broke a line with my PR last week. I fixed it and also resolved the remaining issues and rebased to master. This should be good to go, but please take another look. I asserted that the stress limiter is not combined with VEP for now. I also now understand why the strain rate is handled differently (the x2 only applies to the viscous stress part, but not the elastic stress part, which is correctly considered in the visco-elastic part, but not the drucker prager plasticity, which is why the strain rate has to be modified to compute the correct plastic behavior, just documenting here and in the code for future reference).

naliboff commented:
@gassmoeller - Just had another look through and agree everything looks to go. Thank you for the review and last two commits! I'll let you merge when ready.

naliboff commented:
@gassmoeller - Once this is merged, is there any reason to wait on removing `viscoelastic_plastic` and all associated files? The benchmarks using `viscoelastic_plastic` have all been updated here to use `visco_plastic`.

gassmoeller commented:
Ok, I will merge this.

> Once this is merged, is there any reason to wait on removing viscoelastic_plastic and all associated files? 

No, please remove viscoelastic_plastic. Removing duplication is always good.

naliboff commented:
Follow-up: #3439 removes the viscoelastic plastic material model and test suite.
